File: frontend/src/services/wallet.ts
import api from './api';

export interface WalletBalance {
  available_balance: number;
  locked_balance: number;
  total_balance: number;
  currency: string;
}

export interface Transaction {
  id: number;
  user_id: number;
  amount: number;
  transaction_type: 
    | 'deposit' | 'withdrawal' 
    | 'boom_purchase' | 'boom_sell'
    | 'deposit_real' | 'withdrawal_real' | 'boom_purchase_real' | 'boom_sell_real'
    | 'redistribution_received'
    | 'bonus_received' | 'royalties_received' | 'gift_received'
    | 'bonus_payout' | 'royalties_payout' | 'gift_sent'
    | 'bonus_received_real' | 'royalties_received_real' | 'gift_received_real'
    | 'commission_received_real' | 'cashback_real' | 'reward_real'
    | 'fee_real' | 'commission_paid_real' | 'gift_sent_real'
    | 'penalty_real' | 'tax_real'
    | 'royalties_redistribution' | 'community_bonus' | 'system_redistribution';
  description: string;
  status: 'completed' | 'pending' | 'failed';
  created_at: string;
}

export interface DepositRequest {
  amount: number;
  phone_number: string;
}

export interface WithdrawalRequest {
  amount: number;
  phone_number: string;
}

export interface DetailedBalance {
  liquid_balance: number;
  bom_value: number;
  total_balance: number;
  currency: string;
}

export interface DualBalance {
  real_balance: number;
  virtual_balance: number;
  total_balance: number;
  currency: string;
  real_source: string;
  virtual_source: string;
}

export interface CashBalance {
  available: number;
  locked: number;
  usable: number;
  currency: string;
}

export interface VirtualBalance {
  balance: number;
  currency: string;
  source: string;
}

export interface PurchaseCheck {
  canPurchase: boolean;
  availableBalance: number;
  requiredAmount: number;
  missingAmount: number;
  isVirtual: boolean;
}

export interface CompleteBalance {
  cash: CashBalance;
  virtual: VirtualBalance;
  total: number;
  currency: string;
}

export function getTransactionTarget(transactionType: string): 'real' | 'virtual' | 'neutral' {
  const realTypes = [
    'deposit', 'withdrawal', 'boom_purchase', 'boom_sell', 
    'deposit_real', 'withdrawal_real', 'boom_purchase_real', 'boom_sell_real',
    'gift_received', 'bonus_received', 'fee', 'commission',
    'bonus_received_real', 'royalties_received_real', 'gift_received_real',
    'commission_received_real', 'cashback_real', 'reward_real',
    'fee_real', 'commission_paid_real', 'gift_sent_real',
    'penalty_real', 'tax_real'
  ];
  
  const virtualTypes = [
    'redistribution_received',
    'royalties_redistribution', 'community_bonus', 'system_redistribution'
  ];
  
  if (realTypes.includes(transactionType)) {
    return 'real';
  } else if (virtualTypes.includes(transactionType)) {
    return 'virtual';
  } else {
    return 'neutral';
  }
}

class WalletService {
  private cashBalanceCache: CashBalance | null = null;
  private virtualBalanceCache: VirtualBalance | null = null;
  private lastRefresh = 0;
  private readonly CACHE_DURATION = 10000;

  /**
   * 🔴 MÉTHODE AMÉLIORÉE : Rafraîchir le solde CASH
   */
  async refreshCashBalance(force: boolean = false): Promise<CashBalance> {
    try {
      // ✅ AJOUT: ID unique pour tracking des requêtes
      const requestId = Date.now();
      console.log(`🔄 [WALLET-SERVICE] Refresh cash #${requestId} (force: ${force})`);
      
      const now = Date.now();
      const shouldRefresh = force || 
                           !this.cashBalanceCache || 
                           (now - this.lastRefresh) > this.CACHE_DURATION;

      if (!shouldRefresh && this.cashBalanceCache) {
        console.log(`💰 [WALLET-SERVICE] Utilisation du cache cash balance (req #${requestId})`);
        return this.cashBalanceCache;
      }

      console.log(`🔄 [WALLET-SERVICE] refreshCashBalance: Appel API (req #${requestId})`);
      
      // 🔴 AMÉLIORATION: Invalider le cache côté client
      const params: any = { _t: Date.now() };
      if (force) {
        params.force = true;
      }
      
      const response = await api.get('/wallet/cash-balance', { params });

      const cashData = response.data;
      const cashBalance: CashBalance = {
        available: cashData.balance || cashData.available_balance || 0,
        locked: cashData.locked_balance || 0,
        usable: (cashData.balance || 0) - (cashData.locked_balance || 0),
        currency: cashData.currency || 'FCFA'
      };

      this.cashBalanceCache = cashBalance;
      this.lastRefresh = now;

      console.log(`✅ [WALLET] refreshCashBalance: Nouveau solde ${cashBalance.available} ${cashBalance.currency}`);
      
      return cashBalance;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur refreshCashBalance:', error);
      
      if (this.cashBalanceCache) {
        console.warn('⚠️ [WALLET] Utilisation du cache suite à erreur API');
        return this.cashBalanceCache;
      }
      
      return { 
        available: 0, 
        locked: 0, 
        usable: 0, 
        currency: 'FCFA' 
      };
    }
  }

  /**
   * 🔴 MÉTHODE AMÉLIORÉE : Rafraîchir le solde VIRTUEL
   */
  async refreshVirtualBalance(force: boolean = false): Promise<VirtualBalance> {
    try {
      const now = Date.now();
      const shouldRefresh = force || 
                           !this.virtualBalanceCache || 
                           (now - this.lastRefresh) > this.CACHE_DURATION;

      if (!shouldRefresh && this.virtualBalanceCache) {
        console.log('🎁 [WALLET] Utilisation du cache virtual balance');
        return this.virtualBalanceCache;
      }

      console.log('🔄 [WALLET] refreshVirtualBalance: Appel API pour user');
      
      // 🔴 AMÉLIORATION: Invalider le cache côté client
      const params: any = { _t: Date.now() };
      if (force) {
        params.force = true;
      }
      
      const response = await api.get('/wallet/balance', { params });

      const virtualData = response.data;
      const virtualBalance: VirtualBalance = {
        balance: virtualData.balance || virtualData.available_balance || 0,
        currency: virtualData.currency || 'FCFA',
        source: 'Redistributions communautaires'
      };

      this.virtualBalanceCache = virtualBalance;
      this.lastRefresh = now;

      console.log(`✅ [WALLET] refreshVirtualBalance: Nouveau solde ${virtualBalance.balance} ${virtualBalance.currency}`);
      
      return virtualBalance;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur refreshVirtualBalance:', error);
      
      if (this.virtualBalanceCache) {
        console.warn('⚠️ [WALLET] Utilisation du cache suite à erreur API');
        return this.virtualBalanceCache;
      }
      
      return { 
        balance: 0, 
        currency: 'FCFA', 
        source: 'Système' 
      };
    }
  }

  /**
   * 🔴 MÉTHODE AMÉLIORÉE : Rafraîchir TOUS les soldes
   */
  async refreshAllBalances(force: boolean = false): Promise<CompleteBalance> {
    try {
      const [cash, virtual] = await Promise.all([
        this.refreshCashBalance(force),
        this.refreshVirtualBalance(force)
      ]);

      return {
        cash,
        virtual,
        total: cash.usable + virtual.balance,
        currency: cash.currency
      };
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur refreshAllBalances:', error);
      return {
        cash: { available: 0, locked: 0, usable: 0, currency: 'FCFA' },
        virtual: { balance: 0, currency: 'FCFA', source: 'Erreur' },
        total: 0,
        currency: 'FCFA'
      };
    }
  }

  /**
   * Rafraîchir les transactions
   */
  async refreshTransactions(): Promise<Transaction[]> {
    try {
      const response = await api.get('/wallet/transactions', {
        params: { force: true }
      });
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur refreshTransactions:', error);
      return [];
    }
  }

  /**
   * Obtenir le solde détaillé
   */
  async getDetailedBalance(): Promise<DetailedBalance> {
    try {
      const response = await api.get('/wallet/balance/detailed');
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur solde détaillé:', error);
      return {
        liquid_balance: 0,
        bom_value: 0,
        total_balance: 0,
        currency: 'FCFA'
      };
    }
  }

  /**
   * Obtenir le solde simple (VIRTUEL - Wallet)
   */
  async getBalance(): Promise<WalletBalance> {
    try {
      const response = await api.get('/wallet/balance');
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur solde:', error);
      return {
        available_balance: 0,
        locked_balance: 0,
        total_balance: 0,
        currency: 'FCFA'
      };
    }
  }

  /**
   * Obtenir le solde RÉEL (CashBalance) - POUR ACHATS BOOM
   */
  async getRealBalance(): Promise<{ balance: number; currency: string }> {
    try {
      const response = await api.get('/wallet/cash-balance');
      return {
        balance: response.data.balance || 0,
        currency: response.data.currency || 'FCFA'
      };
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur argent réel:', error);
      return { balance: 0, currency: 'FCFA' };
    }
  }

  /**
   * Obtenir les DEUX soldes en une seule requête
   */
  async getDualBalance(): Promise<DualBalance> {
    try {
      const response = await api.get('/wallet/dual-balance');
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur soldes duals:', error);
      return {
        real_balance: 0,
        virtual_balance: 0,
        total_balance: 0,
        currency: 'FCFA',
        real_source: 'CashBalance - Pour achats BOOM',
        virtual_source: 'Wallet - Bonus & redistributions'
      };
    }
  }

  /**
   * Obtenir l'historique des transactions
   */
  async getTransactions(): Promise<Transaction[]> {
    try {
      const response = await api.get('/wallet/transactions');
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur transactions:', error);
      return [];
    }
  }

  /**
   * Déposer des fonds RÉELS
   */
  async deposit(depositData: DepositRequest): Promise<Transaction> {
    try {
      const response = await api.post('/wallet/deposit', depositData);
      
      // 🔴 AMÉLIORATION: Refresh forcé après dépôt
      await this.refreshCashBalance(true);
      
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur dépôt:', error);
      throw error;
    }
  }

  /**
   * Retirer des fonds RÉELS
   */
  async withdraw(withdrawalData: WithdrawalRequest): Promise<Transaction> {
    try {
      const response = await api.post('/wallet/withdraw', withdrawalData);
      
      // 🔴 AMÉLIORATION: Refresh forcé après retrait
      await this.refreshCashBalance(true);
      
      return response.data;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur retrait:', error);
      throw error;
    }
  }

  /**
   * Obtenir uniquement le solde RÉEL (pour affichage clair)
   */
  async getCashBalance(): Promise<CashBalance> {
    return this.refreshCashBalance(false);
  }

  /**
   * Obtenir uniquement le solde VIRTUEL (redistributions)
   */
  async getVirtualBalance(): Promise<VirtualBalance> {
    return this.refreshVirtualBalance(false);
  }

  /**
   * Vérifier si un achat est possible (solde réel suffisant)
   */
  async canPurchase(amount: number): Promise<PurchaseCheck> {
    try {
      const cashBalance = await this.getCashBalance();
      const canBuy = cashBalance.usable >= amount;
      
      return {
        canPurchase: canBuy,
        availableBalance: cashBalance.usable,
        requiredAmount: amount,
        missingAmount: Math.max(0, amount - cashBalance.usable),
        isVirtual: false
      };
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur vérification achat:', error);
      return {
        canPurchase: false,
        availableBalance: 0,
        requiredAmount: amount,
        missingAmount: amount,
        isVirtual: false
      };
    }
  }

  /**
   * Obtenir le détail complet des soldes (pour debug)
   */
  async getCompleteBalance(): Promise<CompleteBalance> {
    return this.refreshAllBalances(false);
  }

  /**
   * Obtenir le montant bloqué (si applicable)
   */
  async getLockedBalance(): Promise<number> {
    try {
      const response = await api.get('/wallet/locked-balance');
      return response.data.locked_balance || 0;
    } catch (error: any) {
      console.error('❌ [WALLET] Erreur montant bloqué:', error);
      return 0;
    }
  }

  /**
   * Invalider le cache (après une action utilisateur)
   */
  invalidateCache(): void {
    this.cashBalanceCache = null;
    this.virtualBalanceCache = null;
    this.lastRefresh = 0;
    console.log('🗑️ [WALLET] Cache invalidé');
  }
}

// Export unique
export const walletService = new WalletService();

File: backend/app/routes/wallet.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from decimal import Decimal
from app.database import get_db
from app.models.user_models import User, Wallet
from app.models.payment_models import CashBalance
from app.schemas.wallet_schemas import (
    WalletBalance, TransactionResponse, DepositRequest, WithdrawalRequest
)
from app.services.wallet_service import (
    get_wallet_balance, get_transaction_history, create_transaction, has_sufficient_funds
)
from app.services.auth import get_current_user_from_token as get_current_user
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/wallet", tags=["wallet"])

@router.get("/balance", response_model=WalletBalance)
def get_balance(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Récupérer le solde du portefeuille VIRTUEL (Wallet)"""
    try:
        return get_wallet_balance(db, current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/cash-balance", response_model=WalletBalance)
def get_cash_balance(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Récupérer le solde RÉEL (CashBalance) - pour achats BOOM"""
    try:
        cash_balance = db.query(CashBalance).filter(
            CashBalance.user_id == current_user.id
        ).first()
        
        if not cash_balance:
            cash_balance = CashBalance(
                user_id=current_user.id,
                available_balance=Decimal('0.00'),
                currency="FCFA"
            )
            db.add(cash_balance)
            db.commit()
            db.refresh(cash_balance)
        
        return WalletBalance(
            balance=float(cash_balance.available_balance),
            currency=cash_balance.currency
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Erreur récupération CashBalance: {str(e)}")

# ✅ NOUVELLE ROUTE POUR LES DEUX SOLDES
@router.get("/dual-balance")
def get_dual_balance(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Récupérer les DEUX soldes en une seule requête"""
    try:
        # 1. Argent RÉEL (CashBalance)
        cash_balance = db.query(CashBalance).filter(
            CashBalance.user_id == current_user.id
        ).first()
        
        real_balance = Decimal('0.00')
        if cash_balance and cash_balance.available_balance:
            real_balance = cash_balance.available_balance
        
        # 2. Argent VIRTUEL (Wallet)
        wallet_data = get_wallet_balance(db, current_user.id)
        virtual_balance = Decimal(str(wallet_data.get("balance", 0)))
        
        return {
            "real_balance": float(real_balance),
            "virtual_balance": float(virtual_balance),
            "total_balance": float(real_balance + virtual_balance),
            "currency": "FCFA",
            "real_source": "CashBalance - Pour achats BOOM",
            "virtual_source": "Wallet - Bonus & redistributions"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Erreur récupération soldes: {str(e)}")

@router.get("/check-real-funds/{amount}")
def check_real_funds(
    amount: float,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Vérifier spécifiquement les fonds RÉELS (pour achats BOOM)
    """
    try:
        # Utiliser la fonction modifiée avec fund_type="real"
        from app.services.wallet_service import has_sufficient_funds
        result = has_sufficient_funds(
            db=db,
            user_id=current_user.id,
            amount=amount,
            lock_for_check=False,
            fund_type="real"
        )
        
        return {
            "can_purchase": result["has_funds"],
            "available_real_balance": float(Decimal(result["balance"])),
            "required_amount": amount,
            "missing": float(Decimal(result["missing"])),
            "source": result.get("source", "CashBalance")
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Erreur vérification fonds: {str(e)}")

@router.get("/transactions", response_model=list[TransactionResponse])
def get_transactions(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Récupérer l'historique des transactions"""
    try:
        return get_transaction_history(db, current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/deposit", response_model=TransactionResponse)
def deposit_funds(
    deposit_data: DepositRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Déposer de l'argent RÉEL (Mobile Money -> CashBalance)"""
    try:
        if deposit_data.amount <= 0:
            raise HTTPException(status_code=400, detail="Le montant doit être positif")
        
        # 1. Vérifier/Créer CashBalance
        cash_balance = db.query(CashBalance).filter(
            CashBalance.user_id == current_user.id
        ).first()
        
        if not cash_balance:
            cash_balance = CashBalance(
                user_id=current_user.id,
                available_balance=Decimal('0.00'),
                currency="FCFA"
            )
            db.add(cash_balance)
            db.commit()
            db.refresh(cash_balance)
        
        # 2. Mettre à jour CashBalance
        old_balance = cash_balance.available_balance or Decimal('0.00')
        cash_balance.available_balance = old_balance + Decimal(str(deposit_data.amount))
        
        # 3. Créer transaction
        transaction = create_transaction(
            db=db,
            user_id=current_user.id,
            amount=deposit_data.amount,
            transaction_type="deposit_real",
            description=f"Dépôt Mobile Money - {deposit_data.phone_number}"
        )
        
        db.commit()
        
        logger.info(f"✅ Dépôt réussi: {old_balance} → {cash_balance.available_balance}")
        
        return transaction
        
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Erreur lors du dépôt: {str(e)}")

@router.post("/withdraw", response_model=TransactionResponse)
def withdraw_funds(
    withdrawal_data: WithdrawalRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Retirer de l'argent RÉEL (CashBalance -> Mobile Money)"""
    try:
        if withdrawal_data.amount <= 0:
            raise HTTPException(status_code=400, detail="Le montant doit être positif")
        
        # 1. Vérifier CashBalance
        cash_balance = db.query(CashBalance).filter(
            CashBalance.user_id == current_user.id
        ).first()
        
        if not cash_balance:
            raise HTTPException(status_code=400, detail="Aucun solde réel disponible")
        
        available_balance = cash_balance.available_balance or Decimal('0.00')
        withdrawal_amount = Decimal(str(withdrawal_data.amount))
        
        # 2. Vérifier solde suffisant
        if available_balance < withdrawal_amount:
            raise HTTPException(
                status_code=400, 
                detail=f"Solde réel insuffisant: {available_balance} FCFA < {withdrawal_amount} FCFA"
            )
        
        # 3. Débiter CashBalance
        old_balance = available_balance
        cash_balance.available_balance = old_balance - withdrawal_amount
        
        # 4. Créer transaction
        transaction = create_transaction(
            db=db,
            user_id=current_user.id,
            amount=withdrawal_data.amount,
            transaction_type="withdrawal_real",
            description=f"Retrait vers {withdrawal_data.phone_number}"
        )
        
        db.commit()
        
        logger.info(f"✅ Retrait réussi: {old_balance} → {cash_balance.available_balance}")
        
        return transaction
        
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Erreur lors du retrait: {str(e)}")